File: app/src/main/java/com/tommasov/mg4swipenovalauncher/AccService.java
package com.tommasov.mg4swipenovalauncher;

import android.os.Build;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import android.accessibilityservice.AccessibilityService;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.view.accessibility.AccessibilityEvent;

public class AccService extends AccessibilityService {

    @Override
    public void onCreate() {
        super.onCreate();
        IntentFilter filter = new IntentFilter();
        filter.addAction("com.tommasov.mg4swipenovalauncher.ACTION_BACK");
        LocalBroadcastManager.getInstance(this).registerReceiver(backReceiver, filter);
    }

    private final BroadcastReceiver backReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            performBackAction();
        }
    };

    @Override
    public void onAccessibilityEvent(AccessibilityEvent event) {
    }

    @Override
    public void onInterrupt() {
    }

    public void performBackAction() {
        try { performGlobalAction(GLOBAL_ACTION_BACK); } catch (Exception ignored) {}
    }

    @Override
    public void onDestroy() {
        try { LocalBroadcastManager.getInstance(this).unregisterReceiver(backReceiver); } catch (Exception ignored) {}
        super.onDestroy();
    }
}

File: app/src/main/java/com/tommasov/mg4swipenovalauncher/AccService.kt
package com.tommasov.mg4swipenovalauncher

import android.accessibilityservice.AccessibilityService
import android.view.accessibility.AccessibilityEvent

/**
 * Accessibility service that performs GLOBAL_ACTION_BACK when triggered.
 * Uses a static callback — no broadcasts, no security risk.
 */
class AccService : AccessibilityService() {

    override fun onCreate() {
        super.onCreate()
        instance = this
        AppLogger.i("AccService started")
    }

    override fun onAccessibilityEvent(event: AccessibilityEvent?) {}
    override fun onInterrupt() {}

    override fun onDestroy() {
        instance = null
        AppLogger.i("AccService destroyed")
        super.onDestroy()
    }

    companion object {
        @Volatile private var instance: AccService? = null

        /** Called by SwipeService to trigger the Back action */
        fun triggerBack() {
            val svc = instance
            if (svc != null) {
                try { svc.performGlobalAction(GLOBAL_ACTION_BACK) }
                catch (e: Exception) { AppLogger.e("Back action failed", e) }
            } else {
                AppLogger.w("AccService not running — cannot perform Back")
            }
        }
    }
}
